packages/core/src/rules/oas3/remove-unused-components.ts

```diff
@@ -65,10 +66,13 @@ export const RemoveUnusedComponents: Oas3Rule = () => {
       },
     },
     NamedSchemas: {
-      Schema(schema, { location, key }) {
-        if (!schema.allOf) {
-          registerComponent(location, 'schemas', key.toString());
-        }
+      Schema(schema, { location, key, resolve }) {
+        registerComponent(
+          location,
+          'schemas',
+          key.toString(),
+          schema.allOf && schema.allOf.some((v) => v.$ref && resolve(v).node?.discriminator)
```

Comment:
I'm afraid this won't work correctly with transient allOf-discriminator pair. Consider this example:

```yaml
openapi: '3.0.0'
info: 
  title: X
  description: X
  version: 1.0.0

components: 
  schemas: 
    BaseVehicle: 
      type: object
      description: Vehicle
      discriminator:
        propertyName: powerSource
      properties:
        powerSource:
          description: How is the vehicle powered.
          type: string
          example: pedaling

    Vehicle:
      allOf: 
        - $ref: "#/components/schemas/BaseVehicle"
        - type: object 
          properties:
            vehicleType:
              description: The type of vehicle.
              type: string
              example: bicycle

    PedaledVehicle:
      allOf:
        - $ref: "#/components/schemas/Vehicle"
        - type: object
          description: Pedaled Vehicle
          properties:
            range:
              description: The 95th percentile range of a trip in kilometers.
              type: integer
              example: 100
```

The `PedaledVehicle` will be considered unused despite `Vehicle` resolves to `BaseVehicle` which actually contains the `discriminator`. I'm not sure this is a common use case but technically it looks like a valid one.

Reply:
@tatomyr Appreciate the feedback! And yes, I suppose that is valid in the general case.

I'm wondering if this is fruitless -- I am trying to apply a very strict definition of "used" that works for my use case (in this example, I would actually expect all those schemas to be removed) but wouldn't be generally applicable. 

I can look at taking another run at this (although it might be a few days) but I am starting to think a different approach is necessary.